cmd/check/main.go

```diff
@@ -49,66 +46,259 @@ func main() {
 		if err != nil {
 			logrus.Errorf("could not resolve registry: %s", err)
 			os.Exit(1)
 			return
 		}
 
+		repo, err := name.NewRepository(req.Source.Repository)
+		if err != nil {
+			logrus.Errorf("could not resolve repository: %s", err)
+			os.Exit(1)
+			return
+		}
+
 		repo.Registry = mirror
-		*tag = repo.Tag(req.Source.Tag())
 
-		response, err = checkWithRetry(req.Source.RegistryMirror.BasicCredentials, req.Version, *tag)
+		copy := req.Source
+		copy.Repository = repo.String()
+		copy.BasicCredentials = req.Source.RegistryMirror.BasicCredentials
+		copy.RegistryMirror = nil
+
+		mirrorSource = &copy
+	}
+
+	var response resource.CheckResponse
+
+	if mirrorSource != nil {
+		response, err = checkWithRetry(*mirrorSource, req.Version)
 		if err != nil {
-			logrus.Warnf("checking mirror %s failed: %s", mirror.RegistryStr(), err)
+			logrus.Warnf("checking mirror %s failed: %s", mirrorSource.Repository, err)
 		} else if len(response) == 0 {
-			logrus.Warnf("checking mirror %s failed: tag not found", mirror.RegistryStr())
+			logrus.Warnf("checking mirror %s failed: tag not found", mirrorSource.Repository)
 		}
-
-		repo.Registry = origin
 	}
 
 	if len(response) == 0 {
-		*tag = repo.Tag(req.Source.Tag())
-		response, err = checkWithRetry(req.Source.BasicCredentials, req.Version, *tag)
-	}
-	if err != nil {
-		logrus.Errorf("checking origin %s failed: %s", tag.RegistryStr(), err)
-		os.Exit(1)
-		return
+		response, err = checkWithRetry(req.Source, req.Version)
+		if err != nil {
+			logrus.Errorf("checking origin %s failed: %s", req.Source.Repository, err)
+			os.Exit(1)
+			return
+		}
 	}
 
 	json.NewEncoder(os.Stdout).Encode(response)
 }
 
-func checkWithRetry(principal resource.BasicCredentials, version *resource.Version, ref name.Tag) (CheckResponse, error) {
-	var response CheckResponse
+func checkWithRetry(source resource.Source, version *resource.Version) (resource.CheckResponse, error) {
+	var response resource.CheckResponse
 	err := resource.RetryOnRateLimit(func() error {
 		var err error
-		response, err = check(principal, version, ref)
+		if source.Tag != "" {
+			response, err = checkTag(source, version)
+		} else {
+			response, err = checkRepository(source, version)
+		}
 		return err
 	})
 	return response, err
 }
 
-func check(principal resource.BasicCredentials, version *resource.Version, ref name.Tag) (CheckResponse, error) {
+func checkRepository(source resource.Source, from *resource.Version) (resource.CheckResponse, error) {
+	repo, err := name.NewRepository(source.Repository, name.WeakValidation)
+	if err != nil {
+		return resource.CheckResponse{}, fmt.Errorf("resolve repository: %w", err)
+	}
+
+	auth := &authn.Basic{
+		Username: source.Username,
+		Password: source.Password,
+	}
+
+	imageOpts := []remote.Option{}
+
+	rt, err := transport.New(repo.Registry, auth, http.DefaultTransport, []string{repo.Scope(transport.PullScope)})
```

Comment:
is this required for `remote.List`  and therefore isn't used in checkTag ?

Reply:
It isn't as important for `checkTag` because we'll only make one digest request (or two if the digest changed) in that checking mode. This is only here because the semver tag checking involves one digest request per semver tag. This commit provides some explanation: https://github.com/concourse/registry-image-resource/pull/214/commits/611afb71cf7091f6656d1a10859c1a7e79691b13

I could consolidate them anyway if not just to DRY up the code as you suggested. The difference would be that instead of making an extra auth token request we'd make an extra ping request, which is still better.